stdlib/seq.mli

```
@@ -13,97 +13,807 @@
+(** Provided the function [eq] defines an equality on elements,
+    [uniq eq xs] is the sequence [xs],
+    deprived of its adjacent duplicate elements.
+
+    Thus, [uniq eq xs] has no adjacent duplicate elements.
+    Furthermore, if [xs] is sorted according to an ordering
+    that is compatible with [eq],
+    then [uniq eq xs] has no duplicate elements at all.
+
+    [uniq eq xs] is equivalent to [map head (group eq xs)].
+
+    @since 4.14 *)
+
+val group : ('a -> 'a -> bool) -> 'a t -> 'a t t
+(** Provided the function [eq] defines an equality on elements,
+    [group eq xs] is the sequence of the maximal runs
+    of adjacent duplicate elements of the sequence [xs].
+
+    Every element of [group eq xs] is a nonempty sequence of equal elements.
+
+    The concatenation [concat (group eq xs)] is equal to [xs].
+
+    Consuming [group eq xs], and consuming the sequences that it contains,
+    can cause [xs] to be consumed more than once. Therefore, [xs] must be
+    persistent.
+
+    @since 4.14 *)
+
+val memoize : 'a t -> 'a t
+(** The sequence [memoize xs] has the same elements as the sequence [xs].
+
+    Regardless of whether [xs] is ephemeral or persistent,
+    [memoize xs] is persistent: even if it is queried several times,
+    [xs] is queried at most once.
+    @since 4.14 *)
+
+exception ForcedTwice
+(** This exception is raised when a sequence returned by [once]
+    (or a suffix of it) is queried more than once.
+    @since 4.14 *)
+
+val once: 'a t -> 'a t
+(** The sequence [once xs] has the same elements as the sequence [xs].
+
+    Regardless of whether [xs] is ephemeral or persistent,
+    [once xs] is an ephemeral sequence: it can be queried at most once.
+    If it (or a suffix of it) is queried more than once, then the exception
+    [ForcedTwice] is raised. This can be useful, while debugging or testing,
+    to ensure that a sequence is consumed at most once.
+
+    @raise ForcedTwice if [once xs], or a suffix of it,
```

Do you feel a new exception is warranted ? That is is the exception made to be caught ? If not then you could also simply raise `Invalid_argument` (i.e. programming error). Just asking the question I'm not using `Seq` enough to have an opinion on this.

---

not sure, but Invalid_argument seems less specific than something custom. It's not intended to be caught.

---

If it's not intended to be caught, then it's a programming error. For which we already have `Invalid_argument` as a convention.

---

There is no good reason to use `once` in production, I think. That said, while debugging, `once` can be useful, and catching `Forced_twice` might be useful. I prefer not to re-use `Invalid_argument`, whose name is really not appropriate here; the problem that is detected is not an invalid argument, but an invalid use.